scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java

```diff
@@ -1004,26 +1005,49 @@ public final void handshakeCompleted() {
 	}
 
 	/**
-	 * Notifies all registered session listeners about a handshake
-	 * failure.
+	 * Notifies all registered session listeners about a handshake failure.
+	 * 
+	 * If {@link #setFailureCause(Throwable)} was called before, only calls with
+	 * the same cause will notify the listeners.
 	 * 
 	 * @param cause The reason for the failure.
 	 */
 	public final void handshakeFailed(Throwable cause) {
-		if (!handshakeFailed) {
+		if (this.cause == null) {
+			this.cause = cause;
+		}
+		if (!handshakeFailed && this.cause == cause) {
 			handshakeFailed = true;
 			setPendingFlight(null);
 			if (!sessionEstablished) {
 				for (SessionListener sessionListener : sessionListeners) {
 					sessionListener.handshakeFailed(this, cause);
 				}
-				for (RawData message : takeDeferredApplicationData()) {
-					message.onError(cause);
-				}
 			}
 		}
 	}
 
+	/**
+	 * Get cause of failure.
+	 * 
+	 * @return cause of failure, or {@code null}, if handshake has failed.
+	 */
+	public Throwable getFailureCause() {
+		return cause;
+	}
+
+	/**
+	 * Set the failure cause.
+	 * 
+	 * Prevent to report secondary causes when executing the cleanup.
```

Comment:
What are you calling "executing the cleaning" ? (on flight retransmission timeout ?)

Reply:
The main cause of this PR was the `Connection.updatePeerAddress`. Always, if `terminate`/`remove` was called before `handshakeFailed`, the "secondary cause, address changed" was reported.
On pain point is a unit test, which used "handshakeFailed" to check, if the connextion store has removed the connection. With that, "remove" must be called before the "handshakeFailed". Therefore I decided to introduce the "failure cause".

Reply:
I'm lost :sweat_smile: 

I think I get the point about `updatePeerAddress` which always implies that handshake reports "address changed". 

But I don't understand the link with unit tests ? 

I think `Prevent to report secondary causes when executing the cleanup` javadoc message is a bit confusing. Even if this is the reason why you introduce this. At `Handshake` level this setter is just a new feature. (The real workaround is in `Connection.updatePeerAddress` and if you care about explain it, you should explain it in `updatePeerAddress`)

Reply:
`DTLSConnectorAdvancedTest.LatchSessionListener`

```
Throwable error = clientSessionListener.waitForSessionFailed(timeout, TimeUnit.MILLISECONDS);
assertNotNull("client handshake not failed", error);
assertThat(clientConnectionStore.remainingCapacity(), is(remain));
```

That assumes, that the connection is removed before the `handshakeFailed` is called.

Reply:
> I think Prevent to report secondary causes when executing the cleanup javadoc message is a bit confusing. Even if this is the reason why you introduce this. At Handshake level this setter is just a new feature. (The real workaround is in Connection.updatePeerAddress and if you care about explain it, you should explain it in updatePeerAddress)

No the real, implementation is in `Handshaker.handshakeFailed`.
Even if the `updatePeerAddress` would not check the cause, it will not trigger the listeners to be called, if the cause is set before. My intention in add that check to the `updatePeerAddress` was to document that behaviour. Now I think, a comment would be better ...

Reply:
It's clearer to me now.
But `Prevent to report secondary causes when executing the cleanup.` javadoc sounds still confusing.
Anyway it's not a big deal.